### PR TITLE
Update crypto key search to find keys between blocks ##search

### PR DESCRIPTION
**Pull request by sylvainpelissier**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
Previously, the search failed if the key was in between two blocks. The search functions have also been integrated like other search function and are not anymore special cases in **do_string_search** function.

**Test plan**
One test have been added.

**Closing issues**

None.

...

### CODE REVIEW

File: libr/core/cmd_search.c [hunk 1]
```diff
@@ -31,7 +31,7 @@ static const char *help_msg_slash_m[] = {
 
 static const char *help_msg_slash[] = {
 	"Usage:", "/[!bf] [arg]", "Search stuff (see 'e??search' for options)\n"
-	"|Use io.va for searching in non virtual addressing spaces",
+				  "|Use io.va for searching in non virtual addressing spaces",
```
Comment (trufae) [inline libr/core/cmd_search.c:34]: wat, remove those tabs

File: libr/core/cmd_search.c [hunk 2]
```diff
@@ -2399,10 +2399,10 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
 			}
 
 			const ut64 from = itv.addr, to = r_itv_end (itv),
-					from1 = search->bckwrds ? to : from,
-					to1 = search->bckwrds ? from : to;
+				   from1 = search->bckwrds? to: from,
```
Comment (trufae) [inline libr/core/cmd_search.c:2402]: use tabs

File: libr/core/cmd_search.c [hunk 3]
```diff
@@ -2422,23 +2422,21 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
...
-						break;
-					}
-				} else {
-					(void)r_search_update (core->search, at, buf, len);
-					if (core->search->maxhits > 0 && core->search->nhits >= core->search->maxhits) {
-						goto done;
+				} else if (param->privkey_search) {
+					// Adjust length to search between blocks.
+					if (len == core->blocksize) {
+						len = len - 10;
```
Comment (trufae) [inline libr/core/cmd_search.c:2434]: len -= 10;

File: libr/core/cmd_search.c [hunk 4]
```diff
@@ -2422,23 +2422,21 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
...
-					int t = 0;
-					if (param->aes_search) {
-						t = r_search_aes_update (core->search, at, buf, len);
-					} else if (param->privkey_search) {
-						t = r_search_privkey_update (core->search, at, buf, len);
+				r_search_update (core->search, at, buf, len);
+				if (param->aes_search) {
+					// Adjust length to search between blocks.
+					if (len == core->blocksize) {
+						len = len - 39;
```
Comment (trufae) [inline libr/core/cmd_search.c:2429]: len -= 39

File: libr/search/aes-find.c [hunk 5]
```diff
@@ -49,31 +49,53 @@ static bool aes128_key_test(const unsigned char *buf) {
...
+	const int old_nhits = s->nhits;
 
 	r_list_foreach (s->kws, iter, kw) {
 		if (last > 0) {
 			for (i = 0; i < last; i++) {
 				if (aes128_key_test (buf + i)) {
 					kw->keyword_length = 16;
-					return r_search_hit_new (s, kw, from + i);
+					t = r_search_hit_new (s, kw, from + i);
+					i = i + 24;
```
Comment (trufae) [inline libr/search/aes-find.c:63]: i += 24

**Comment by sylvainpelissier** [comment]
It sould be ok now.

File: libr/core/cmd_search.c [hunk 6]
```diff
@@ -2422,23 +2422,21 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
...
-					int t = 0;
-					if (param->aes_search) {
-						t = r_search_aes_update (core->search, at, buf, len);
-					} else if (param->privkey_search) {
-						t = r_search_privkey_update (core->search, at, buf, len);
+				r_search_update (core->search, at, buf, len);
+				if (param->aes_search) {
+					// Adjust length to search between blocks.
+					if (len == core->blocksize) {
+						len -= 39;
```
Comment (ret2libc) [inline libr/core/cmd_search.c:2429]: where does this `39` come from?

File: libr/core/cmd_search.c [hunk 7]
```diff
@@ -2422,23 +2422,21 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
...
-						break;
-					}
-				} else {
-					(void)r_search_update (core->search, at, buf, len);
-					if (core->search->maxhits > 0 && core->search->nhits >= core->search->maxhits) {
-						goto done;
+				} else if (param->privkey_search) {
+					// Adjust length to search between blocks.
+					if (len == core->blocksize) {
+						len -= 10;
```
Comment (ret2libc) [inline libr/core/cmd_search.c:2434]: also this `10`?

File: libr/core/cmd_search.c [hunk 8]
```diff
@@ -3343,6 +3341,9 @@ static int cmd_search(void *data, const char *input) {
 			{
 				RSearchKeyword *kw;
 				kw = r_search_keyword_new_hexmask ("00", NULL);
+				// AES search is at most 40 bytes
+				kw->keyword_length = 40;
```
Comment (ret2libc) [inline libr/core/cmd_search.c:3345]: if there are not already, please define these constant values somewhere appropriate and use a meaningful name instead of hardcoding values like `40` or `14`.

File: libr/search/aes-find.c [hunk 9]
```diff
@@ -49,31 +49,53 @@ static bool aes128_key_test(const unsigned char *buf) {
 }
 
 R_API int r_search_aes_update(RSearch *s, ut64 from, const ut8 *buf, int len) {
-	int i, last = len - 20;
+	int i, t, last = len - 23;
```
Comment (ret2libc) [inline libr/search/aes-find.c:52]: i have no idea why `- 23`. Also, what happens if `len < 23`?

File: libr/search/aes-find.c [hunk 10]
```diff
@@ -49,31 +49,53 @@ static bool aes128_key_test(const unsigned char *buf) {
...
+					i += 24;
+					if (!t) {
+						return -1;
+					}
+					if (t > 1) {
+						return s->nhits - old_nhits;
+					}
 				}
-				if (len - i - 28 > 0) {
+				if (len - i - 31 > 0) {
```
Comment (ret2libc) [inline libr/search/aes-find.c:71]: other hardcoded numbers

Comment (ret2libc) [reply libr/search/aes-find.c:71]: I think you can do directly `if (len - i - SOME_CONSTANT > 0 && aes192_key_test (buf + i)) {`

**Comment by sylvainpelissier** [comment]
Thanks @ret2libc for you comments. I integrated them.

File: libr/search/aes-find.c [hunk 11]
```diff
@@ -49,31 +53,51 @@ static bool aes128_key_test(const unsigned char *buf) {
...
+					if (t > 1) {
+						return s->nhits - old_nhits;
+					}
 				}
-				if (len - i - 28 > 0) {
-					if (aes192_key_test (buf + i)) {
-						kw->keyword_length = 24;
-						return r_search_hit_new (s, kw, from + i);
+				if (len - i - AES192_MIN_SEARCH_LENGTH >= 0 && aes192_key_test (buf + i)) {
+					kw->keyword_length = 24;
```
Comment (ret2libc) [inline libr/search/aes-find.c:76]: why do we test for `AES192_MIN_SEARCH_LENGTH` and then do `kw->keyword_length = 24`? Same in the other cases.

Comment (sylvainpelissier) [reply libr/search/aes-find.c:76]: A 192-bit AES key is 24 bytes so we use `kw->keyword_length = 24` as output to print the key enteriely.  AES192_MIN_SEARCH_LENGTH  give the minimum number of bytes you need to perform the search if it is less you cannot search.

File: libr/search/privkey-find.c [hunk 12]
```diff
@@ -71,10 +72,11 @@ R_API int r_search_privkey_update(RSearch *s, ut64 from, const ut8 *buf, int len
...
 	}
 	r_list_foreach (s->kws, iter, kw) {
-		for (i = 2; i < len - sizeof (rsa_versionmarker); i++) {
-			if (memcmp (&buf[i], rsa_versionmarker, sizeof (rsa_versionmarker)) && 
-			memcmp (&buf[i], ecc_versionmarker, sizeof (ecc_versionmarker)) &&
-			memcmp (&buf[i], safecurves_versionmarker, sizeof (safecurves_versionmarker))) {
+		// Iteration until the remaining length is too small to contain a key.
+		for (i = 2; i < len - sizeof (rsa_versionmarker) - 4; i++) {
+			if (memcmp (&buf[i], rsa_versionmarker, sizeof (rsa_versionmarker)) &&
+				memcmp (&buf[i], ecc_versionmarker, sizeof (ecc_versionmarker)) &&
```
Comment (trufae) [inline libr/search/privkey-find.c:78]: use buf +i instead of &buf[i]

File: libr/search/privkey-find.c [hunk 13]
```diff
@@ -60,9 +60,10 @@ static int check_fields(const ut8 *start) {
 // As defined in RFC 3447 for RSA, as defined in RFC 5915 for 
 // elliptic curves and as defined in 7 of RFC 8410 for SafeCurves
 R_API int r_search_privkey_update(RSearch *s, ut64 from, const ut8 *buf, int len) {
-	int i, k, max, index;
+	int i, k, max, index, t;
 	RListIter *iter;
 	RSearchKeyword *kw;
+	const int old_nhits = s->nhits;
```
Comment (trufae) [inline libr/search/privkey-find.c:66]: hits cant be negative, better use size_t

File: libr/search/aes-find.c [hunk 14]
```diff
@@ -49,31 +53,51 @@ static bool aes128_key_test(const unsigned char *buf) {
...
 				}
-				if (len - i - 36 > 0) {
-					if (aes256_key_test (buf + i)) {
-						kw->keyword_length = 32;
-						return r_search_hit_new (s, kw, from + i);
+				if (len - i - AES256_MIN_SEARCH_LENGTH >= 0 && aes256_key_test (buf + i)) {
+					kw->keyword_length = 32;
+					t = r_search_hit_new (s, kw, from + i);
+					i = i + AES256_MIN_SEARCH_LENGTH;
+					if (!t) {
```
Comment (trufae) [inline libr/search/aes-find.c:90]: check this right before the i line

File: libr/search/aes-find.c [hunk 15]
```diff
@@ -49,31 +53,51 @@ static bool aes128_key_test(const unsigned char *buf) {
...
 					}
 				}
-				if (len - i - 36 > 0) {
-					if (aes256_key_test (buf + i)) {
-						kw->keyword_length = 32;
-						return r_search_hit_new (s, kw, from + i);
+				if (len - i - AES256_MIN_SEARCH_LENGTH >= 0 && aes256_key_test (buf + i)) {
+					kw->keyword_length = 32;
+					t = r_search_hit_new (s, kw, from + i);
+					i = i + AES256_MIN_SEARCH_LENGTH;
```
Comment (trufae) [inline libr/search/aes-find.c:89]: use i +=

File: libr/search/aes-find.c [hunk 16]
```diff
@@ -49,31 +53,51 @@ static bool aes128_key_test(const unsigned char *buf) {
...
+					if (!t) {
+						return -1;
+					}
+					if (t > 1) {
+						return s->nhits - old_nhits;
 					}
 				}
 			}
+		} else {
```
Comment (trufae) [inline libr/search/aes-find.c:98]: unnnecessary, there's a return -1 below

File: libr/search/search.c [hunk 17]
```diff
@@ -110,19 +111,20 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
+	// kw->last is used by string search, the right endpoint of last match (forward search), to honor search.overlap
 	kw->last = s->bckwrds ? addr : addr + kw->keyword_length;
+
 	if (s->callback) {
 		int ret = s->callback (kw, s->user, addr);
 		kw->count++;
 		s->nhits++;
 		// If callback returns 0 or larger than 1, forwards it; otherwise returns 2 if search.maxhits is reached
-		return !ret || ret > 1 ? ret : s->maxhits && s->nhits >= s->maxhits ? 2 : 1;
+		return !ret || ret > 1? ret : s->maxhits && s->nhits >= s->maxhits ? 2 : 1;
```
Comment (trufae) [inline libr/search/search.c:127]: see the spacing in the 2nd ternary operator

File: libr/core/cmd_search.c [hunk 18]
```diff
@@ -11,6 +11,12 @@ static int cmd_search(void *data, const char *input);
 
 #define USE_EMULATION 0
 
+// Size needed to search a AES key.
+#define AES_SEARCH_LENGTH 40
+
+// Size meeded to search a private key.
```
Comment (trufae) [inline libr/core/cmd_search.c:17]: meeded is not english :P anyway i think the variable namne is meaningful enough, no need for the comments

Comment (ret2libc) [reply libr/core/cmd_search.c:17]: yeah I agree, the names should be already descriptive enough in this case.

File: libr/core/cmd_search.c [hunk 19]
```diff
@@ -2422,23 +2428,21 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
...
-						t = r_search_aes_update (core->search, at, buf, len);
-					} else if (param->privkey_search) {
-						t = r_search_privkey_update (core->search, at, buf, len);
-					}
-					if (!t || t > 1) {
-						break;
+				r_search_update (core->search, at, buf, len);
+				if (param->aes_search) {
+					// Adjust length to search between blocks.
+					if (len == core->blocksize) {
```
Comment (ret2libc) [inline libr/core/cmd_search.c:2434]: why is this needed? It is not clear to me, sorry. If the code is generic enough, there should be no need to update the len here in different ways based on the kind of search. How it it done for other cases? Can we do the same thing?

File: libr/core/cmd_search.c (hunk 8 above)
Comment (ret2libc) [reply libr/core/cmd_search.c:3351]: `AES_SEARCH_LENGTH`?

File: libr/core/cmd_search.c [hunk 20]
```diff
@@ -3352,6 +3359,9 @@ static int cmd_search(void *data, const char *input) {
 			{
 				RSearchKeyword *kw;
 				kw = r_search_keyword_new_hexmask ("00", NULL);
+				// Private key search is at most 14 bytes
+				kw->keyword_length = 14;
```
Comment (ret2libc) [inline libr/core/cmd_search.c:3363]: above you have `define PRIVATE_KEY_SEARCH_LENGTH 11`, why now this is `14`?

File: libr/search/aes-find.c (hunk 11 above)
Comment (ret2libc) [reply libr/search/aes-find.c:76]: so I guess `AES192_MIN_SEARCH_LENGTH` should actually be named `AES192_KEY_LENGTH` and it should be used both in the line above and here.

File: libr/search/privkey-find.c [hunk 21]
```diff
@@ -71,10 +72,11 @@ R_API int r_search_privkey_update(RSearch *s, ut64 from, const ut8 *buf, int len
 		return -1;
 	}
 	r_list_foreach (s->kws, iter, kw) {
-		for (i = 2; i < len - sizeof (rsa_versionmarker); i++) {
-			if (memcmp (&buf[i], rsa_versionmarker, sizeof (rsa_versionmarker)) && 
-			memcmp (&buf[i], ecc_versionmarker, sizeof (ecc_versionmarker)) &&
-			memcmp (&buf[i], safecurves_versionmarker, sizeof (safecurves_versionmarker))) {
+		// Iteration until the remaining length is too small to contain a key.
+		for (i = 2; i < len - sizeof (rsa_versionmarker) - 4; i++) {
```
Comment (ret2libc) [inline libr/search/privkey-find.c:76]: where is this `4` coming from?

File: libr/core/cmd_search.c [hunk 22]
```diff
@@ -2355,7 +2361,7 @@ static void do_string_search(RCore *core, RInterval search_itv, struct search_pa
 	if (param->inverse) {
 		core->search->maxhits = 1;
 	}
-	if (core->search->n_kws > 0 || param->crypto_search) {
```
[the remark this answers is not in the document]
Comment (sylvainpelissier) [reply libr/core/cmd_search.c:2358]: Thanks, that exactly what I was trying to do and finally I forgot to remove it.

File: libr/core/cmd_search.c (hunk 19 above)
Comment (sylvainpelissier) [reply libr/core/cmd_search.c:2434]: It is needed to search a key even when it is spread over two blocks (defined by the `b` command). I tried to figure out for the other cases how it is handle but I could not understand clearly and I was afraid to break them.

File: libr/search/privkey-find.c (hunk 21 above)
Comment (sylvainpelissier) [reply libr/search/privkey-find.c:76]: I iterate until the remaining length is too small to contain a key.

File: libr/search/aes-find.c (hunk 11 above)
Comment (ret2libc) [reply libr/search/aes-find.c:76]: this ^

File: libr/search/privkey-find.c (hunk 21 above)
Comment (ret2libc) [reply libr/search/privkey-find.c:76]: ok but why `4`? Why not 5, 10 or 0? How did you compute this `4`?

Comment (sylvainpelissier) [reply libr/search/privkey-find.c:76]: After the version number you have the integer tag (one byte) and the length which at least one byte but often three, that's where the 4 comes from.

Comment (ret2libc) [reply libr/search/privkey-find.c:76]: Then there should be a better way to write this.
```
 		for (i = 2; i < len - sizeof (rsa_versionmarker) - PRIVKEY_TAG_SIZE - PRIVKEY_LEN_SIZE; i++) {
```
or something like that. Given you are more familiar with the topic you probably can find better names. Also, this is confusing because at https://github.com/radareorg/radare2/pull/17323/files#diff-3c8c44930d4488e02a4930d2003ba188R71 the function checks for `len < sizeof (rsa_versionmarker)` (and not `len < sizeof (rsa_versionmarker) - PRIVKEY_TAG_SIZE - PRIVKEY_LEN_SIZE`, so it's not even said there are enough bytes to make this subtraction work.

Comment (sylvainpelissier) [reply libr/search/privkey-find.c]: Now I have defined a constant which is the minimal length for a search and I check the length properly.

File: libr/search/aes-find.c (hunk 11 above)
Comment (ret2libc) [reply libr/search/aes-find.c:76]: @sylvainpelissier still this ^

Comment (sylvainpelissier) [reply libr/search/aes-find.c:76]: It was patch in a commit: https://github.com/sylvainpelissier/radare2/blob/02014abbb429bc660aa53842404299c15de888b0/libr/search/aes-find.c#L15

Comment (sylvainpelissier) [reply libr/search/aes-find.c]: I have added them and rebased.

**Comment by sylvainpelissier** [comment]
Everything should be integrated now.

**Comment by XVilka** [comment]
@sylvainpelissier please press "Resolve conversation" when you address review, it helps to see what already had been addressed.

**Comment by ret2libc** [comment]
Sorry I'm still reviewing this :( I'm trying to create some unit tests to help me better understand the API and how they should be used. I still find some aes/privkey specific things a bit weird.

BTW, @sylvainpelissier would you be willing to add unit tests to help test the API and your changes? Not mandatory for this PR, but they usually help and it's very simple to write them. You can look at how they are done by looking at `test/unit/test_io.c`, for example.

File: libr/search/aes-find.c [hunk 23]
```diff
@@ -49,31 +56,49 @@ static bool aes128_key_test(const unsigned char *buf) {
...
 	r_list_foreach (s->kws, iter, kw) {
-		if (last > 0) {
+		if (last >= 0) {
 			for (i = 0; i < last; i++) {
 				if (aes128_key_test (buf + i)) {
-					kw->keyword_length = 16;
-					return r_search_hit_new (s, kw, from + i);
+					kw->keyword_length = AES128_KEY_LENGTH;
+					t = r_search_hit_new (s, kw, from + i);
+					i += AES128_SEARCH_LENGTH;
```
[the remark this answers is not in the document]
Comment (sylvainpelissier) [reply libr/search/aes-find.c]: Ok done.

**Comment by wargio** [comment]
@XVilka you can merge
